### ISSUE TITLE
S3 object metadata not saved when uploaded with presigned url

### ISSUE THREAD

**Issue by kamilbiela** [localstack/localstack#544]
Use case:
I'm enabling users to directly upload to s3 using presigned url. S3 is configured to add event to SQS on Put. Queue consumer, reads the queue and makes HEAD requests with object keys to get the metadata and save information to database (generic image upload, so I know where to add file).

Test script in node js - some ugly code here (to install deps run `npm install aws-sdk request`):
```js
const AWS = require("aws-sdk");
const request = require("request");

let s3 = new AWS.S3({
    endpoint: "http://localhost:4572",
    s3ForcePathStyle: true,
    accessKeyId: "",
    secretAccessKey: "",
    region: "us-west-1"
});

var bucket = "bucketest";
var key = "test.txt";

s3.createBucket({Bucket: bucket}, function (err, data) {
    if (err) {
        console.error(err.message);
        // ignore, probably there is bucket already
    }

    var params = {
        Bucket: bucket,
        Key: key,
        Metadata: {
            venue: "123"
        }
    };

    s3.getSignedUrl('putObject', params, function (err, url) {
        if (err) {
            console.error('Presigning post data encountered an error', err);
        } else {
            console.log('==== URL: ', url);

            var body = new Buffer('Test data.');
            request.put({ url, body, method: "PUT" }, function(err, resp, body) {
                if (err) {
                    console.log('======= error:', error); 
                    return;
                }

                console.log(body);

                s3.headObject({Bucket: bucket, Key: key}, function (err, data) {
                    if (err) console.log("====== error1:", err, err.stack); 
                    else console.log("==== HEAD RESPONSE", data); 
                });
            })
        }
    });
});
```

Output:
```
==== URL:  http://localhost:4572/heaps-test/test.txt?AWSAccessKeyId=somekey&Expires=1515503310&Signature=TgK3B33p2kwCWs5F5KtaZ3fxgXA%3D&x-amz-meta-venue=123
<PutObjectResponse xmlns="http://s3.amazonaws.com/doc/2006-03-01"><PutObjectResponse><ETag>&#34;56dd8a439abf97fda051f88f09f00d65&#34;</ETag><LastModified>2018-01-09T12:53:30.637Z</LastModified></PutObjectResponse></PutObjectResponse>
==== HEAD RESPONSE { LastModified: 2018-01-09T12:53:30.000Z,
  ContentLength: 10,
  ETag: '"56dd8a439abf97fda051f88f09f00d65"',
  ContentType: 'text/html; charset=utf-8',
  Metadata: {} }

```

Expected Output (tested with live AWS): 
```
==== URL:  https://heaps-test.s3.eu-west-1.amazonaws.com/test.txt?AWSAccessKeyId=somekey&Expires=1515503234&Signature=enc17C6glTsVtOiGobugz5NELIc%3D&x-amz-meta-venue=123

==== HEAD RESPONSE { AcceptRanges: 'bytes',
  LastModified: 2018-01-09T12:52:15.000Z,
  ContentLength: 10,
  ETag: '"56dd8a439abf97fda051f88f09f00d65"',
  ContentType: 'binary/octet-stream',
  Metadata: { venue: '123' } }
```

As you can see Metadata is empty when using localstack

**Comment by whummer** [comment]
Thanks for reporting @kamilbiela . This is an upstream issue in `moto`. In particular, the following line in `s3/responses.py` reads metadata only from the HTTP request headers: https://github.com/spulec/moto/blob/master/moto/s3/responses.py#L736

... whereas in your example the metadata needs to be retrieved from the URL query parameters. This is an example request URL:
```
PUT /bucketest/test.txt?AWSAccessKeyId=foo&Expires=1526856841&Signature=abc1rYijcP59W6UHvApY5Vjos%3D&x-amz-meta-venue=123
```

Would you be able to open an issue in https://github.com/spulec/moto , and cross reference this issue? Thanks!

**Comment by sturmer** [comment]
I've opened an issue in moto: https://github.com/spulec/moto/issues/2040
(Does mentioning it in this comment count as cross-referencing it? 🤔 )

**Comment by sturmer** [comment]
I've found something. @whummer: you say that if we use the that URL for PUT, it is supposed to work. But it does! In moto, I've tried:
`        requests.put(url, data={'Bucket': 'mybucket', 'Key': self.name,
                                'Body': self.value, 'Metadata': {'venue': "123"}},
                     headers={'x-amz-meta-venue': "123"})
`

and when then I do `head_object`, I correctly see the metadata. I'm attaching the complete example. Does this mean that this is no longer a valid bug?

**Comment by sturmer** [comment]
```
import boto3
from moto import mock_s3
import requests


class MyModel(object):
    def __init__(self, name, value):
        self.name = name
        self.value = value

    # BROKEN.
    def putMetadataViaHttpPUT(self):
        s3 = boto3.client('s3', region_name='us-east-1')

        url = s3.generate_presigned_url(ClientMethod="put_object", Params={
            'Bucket': 'mybucket', 'Key': 'mykey'})
        requests.put(url, data={'Bucket': 'mybucket', 'Key': self.name,
                                'Body': self.value, 'Metadata': {'venue': "123"}},
                     headers={'x-amz-meta-venue': "123"})

        object_summary = s3.head_object(Bucket='mybucket', Key=self.name)
        metadata = object_summary.get('Metadata')
        # print('Metadata: {}'.format(metadata))
        return metadata

    # Works.
    def putMetadataViaPutObject(self):
        s3 = boto3.client('s3', region_name='us-east-1')

        s3.put_object(Bucket='mybucket', Key=self.name,
                      Body=self.value, Metadata={'venue': "123"})

        object_summary = s3.head_object(Bucket='mybucket', Key=self.name)
        metadata = object_summary.get('Metadata')
        # print('Metadata: {}'.format(metadata))
        return metadata


@mock_s3
def test_my_model_put_object():
    # BROKEN
    conn = boto3.resource('s3', region_name='us-east-1')
    conn.create_bucket(Bucket='mybucket')

    model_instance = MyModel('mykey', 'myvalue')
    return model_instance.putMetadataViaPutObject()


@mock_s3
def test_my_model_http_put():
    # Works.
    conn = boto3.resource('s3', region_name='us-east-1')
    conn.create_bucket(Bucket='mybucket')

    model_instance = MyModel('mykey', 'myvalue')
    return model_instance.putMetadataViaHttpPUT()


# To show the bug, use test_my_model_head_object() to fill metadata. For the working example, use test_my_model_put_object().
if __name__ == '__main__':
    metadata = test_my_model_http_put()
    print('metadata: {}'.format(metadata))
    assert 'venue' in metadata
    print('metadata: {}'.format(metadata))
    assert metadata['venue'] == '123'

**Comment by kamilbiela** [comment]
Hey, if you just put the object the metadata is there, bug happens only when you use presigned url. 

Bug still is there, checked with localstack:latest docker image

**Comment by sturmer** [comment]
Hi @kamilbiela, I'm sorry I misunderstood. You are storing key and metadata already with the `generate_presigned_url` method, and then putting a body.
However, I've retried by putting the metadata in the `get_presigned_url`, and I can see it:
```
import boto3
from moto import mock_s3
import requests


class MyModel(object):
    def __init__(self, name, value):
        self.name = name
        self.value = value

    def putMetadataViaHttpPUT(self):
        s3 = boto3.client('s3', region_name='us-east-1')

        url = s3.generate_presigned_url(ClientMethod="put_object", Params={
            'Bucket': 'mybucket', 'Key': 'mykey', 'Metadata': {'venue': "123"}})

        requests.put(url, data={'Bucket': 'mybucket', 'Key': self.name,
                                'Body': self.value},
                     headers={'x-amz-meta-venue': "123"})

        object_summary = s3.head_object(Bucket='mybucket', Key=self.name)
        return object_summary.get('Metadata')


@mock_s3
def test_my_model_put_object():
    conn = boto3.resource('s3', region_name='us-east-1')
    conn.create_bucket(Bucket='mybucket')

    model_instance = MyModel('mykey', 'myvalue')
    return model_instance.putMetadataViaHttpPUT()


if __name__ == '__main__':
    metadata = test_my_model_put_object()
    assert 'venue' in metadata
    print('metadata: {}'.format(metadata))
    assert metadata['venue'] == '123'
```

The metadata is there. Can you help me understand what I'm doing wrong?

**Comment by sturmer** [comment]
To be clear, I'm claiming something is wrong in localstack, but that moto works fine. Please correct me if I'm wrong!

**Comment by kamilbiela** [comment]
@sturmer Hmm, could be it working because you add extra headers?

```python
        requests.put(url, data={'Bucket': 'mybucket', 'Key': self.name,
                                'Body': self.value},
                     headers={'x-amz-meta-venue': "123"})
```

The url should have the metadata already encoded, and IMO the headers should not be used at all as IRL they come from the user making the upload, so you have no control at them.

**Comment by sturmer** [comment]
D'oh -- now I understand your point @kamilbiela.

Yes, as you say the metadata is there because I've put it in my PUT request. 

The scenario you seem to have an issue with is instead:
1. You create the pre-signed url. **That** incorporates the metadata in the query parameters, silly me 🤦‍♂️ 
2. They upload using PUT, without specifying the metadata.

Thank _you_ for reporting the issue in the first place (and for helping me understand the problem)! I'll try to look into it a bit more :)

**Comment by whummer** [comment]
Thanks for the detailed report @kamilbiela @sturmer @aggansu . This should be fixed in #1745 - can you please give it a try with the latest version.

Closing this issue - please report here if the problem persists. Thanks
